rusoto/services/sts/src/custom/credential.rs

```diff
@@ -89,27 +87,26 @@ impl<T> StsSessionCredentialsClient for T where T: Sts {
 /// To use with MFA, pass in the MFA serial number then set the MFA code.
 /// You will need to ensure the provider has a valid code each time you
 /// acquire a new STS token.
-pub struct StsSessionCredentialsProvider {
-    sts_client: Box<StsSessionCredentialsClient>,
+pub struct StsSessionCredentialsProvider<C: StsSessionCredentialsClient> {
```

Comment:
Rather than this breaking change, perhaps we should implement another struct such as: `StsSessionCredentialsProviderBase` that allows you to pass in a custom sts_client. Then we can make `StsSessionCredentialsProvider` just "wrap" the base provider. This way you can get the ability to specify a custom client, and existing code doesn't break.

Reply:
@SecurityInsanity, thanks for your comment. I attempted to apply your suggestion. However, making StsSessionCredentialsProvider to wrap StsSessionCredentialsProviderBase does not seem to be possible. Here is the code for this attempt (spoiler alert: it does not compile): https://github.com/RelationalAI/rusoto/commit/add524ad1787ee0a5acede81846389eef3548fb4

The other option that comes to my mind is duplicating the code, so StsSessionCredentialsProvider and StsSessionCredentialsProviderBase are not related.

Or, maybe I'm missing another viable option.

Can you please suggest how to proceed?

Reply:
Hey @mdashti I'll take a look later today. Thanks for attempting 😄